client.go

```diff
@@ -142,100 +125,35 @@ func ExchangeConn(c net.Conn, m *Msg) (r *Msg, err error) {
 // It is up to the caller to create a message that allows for larger responses to be
 // returned. Specifically this means adding an EDNS0 OPT RR that will advertise a larger
 // buffer, see SetEdns0. Messages without an OPT RR will fallback to the historic limit
-// of 512 bytes.
-func (c *Client) Exchange(m *Msg, a string) (r *Msg, rtt time.Duration, err error) {
-	return c.ExchangeContext(context.Background(), m, a)
-}
-
-// ExchangeContext acts like Exchange, but honors the deadline on the provided
-// context, if present. If there is both a context deadline and a configured
-// timeout on the client, the earliest of the two takes effect.
-func (c *Client) ExchangeContext(ctx context.Context, m *Msg, a string) (
-	r *Msg,
-	rtt time.Duration,
-	err error) {
+// of 512 bytes
+// To specify a local address or a timeout, the caller has to set the `Client.Dialer`
+// attribute appropriately
+func (c *Client) Exchange(m *Msg, address string) (r *Msg, rtt time.Duration, err error) {
 	if !c.SingleInflight {
-		return c.exchange(ctx, m, a)
+		return c.exchange(m, address)
 	}
-	// This adds a bunch of garbage, TODO(miek).
+
 	t := "nop"
 	if t1, ok := TypeToString[m.Question[0].Qtype]; ok {
 		t = t1
-	}
+	}		
 	cl := "nop"
 	if cl1, ok := ClassToString[m.Question[0].Qclass]; ok {
 		cl = cl1
 	}
 	r, rtt, err, shared := c.group.Do(m.Question[0].Name+t+cl, func() (*Msg, time.Duration, error) {
-		return c.exchange(ctx, m, a)
+		return c.exchange(m, address)
 	})
 	if r != nil && shared {
 		r = r.Copy()
 	}
-	if err != nil {
-		return r, rtt, err
-	}
-	return r, rtt, nil
-}
-
-func (c *Client) dialTimeout() time.Duration {
-	if c.Timeout != 0 {
-		return c.Timeout
-	}
-	if c.DialTimeout != 0 {
-		return c.DialTimeout
-	}
-	return dnsTimeout
-}
-
-func (c *Client) readTimeout() time.Duration {
-	if c.ReadTimeout != 0 {
-		return c.ReadTimeout
-	}
-	return dnsTimeout
-}
-
-func (c *Client) writeTimeout() time.Duration {
-	if c.WriteTimeout != 0 {
-		return c.WriteTimeout
-	}
-	return dnsTimeout
+	return r, rtt, err
 }
 
-func (c *Client) exchange(ctx context.Context, m *Msg, a string) (r *Msg, rtt time.Duration, err error) {
+func (c *Client) exchange(m *Msg, a string) (r *Msg, rtt time.Duration, err error) {
```

Comment:
Why wouldn't we want a context here?

Reply:
If we have a dialer we just don't need a context. The only property of the context used here is the deadline, but there is already a timeout implemented in the dialer in the stdlib.
From the design perspective, using a context would also makes this code more complex, by introducing several different timeouts (context deadline, dialer timeout, default timeout, read/write/dial timeout). A dialer is in my opinion a better place for timeouts, and it is already implemented and handled in the standard library. On the other side, a context carries a deadline, not a timeout (there's quite a difference): a deadline may be already expired when called, while a timeout starts with the request, and can be reused, e.g. for writing and then for reading. More in general Context it's under a lot of discussions on whether it should be used again for what it has been created, i.e. for context-specific data, not for deadlines or other things.
That's my opinion of course, always happy to discuss!